core/meta/src/TStreamerElement.cxx

```diff
@@ -1723,93 +1723,72 @@ TStreamerSTL::TStreamerSTL(const char *name, const char *title, Int_t offset,
    const char *t = trueType;
    if (!t || !*t) t = typeName;
 
-   fTypeName = TClassEdit::ShortType(fTypeName,TClassEdit::kDropStlDefault).c_str();
+   std::string answer;
+   TClassEdit::TSplitType arglist(t, TClassEdit::kDropStlDefault);
+   arglist.ShortType(answer, TClassEdit::kDropStlDefault);
+   fTypeName = answer;
 
    if (name==typeName /* intentional pointer comparison */
        || strcmp(name,typeName)==0) {
       // We have a base class.
       fName = fTypeName;
    }
 
-   Int_t nch = strlen(t);
-   char *s = new char[nch+1];
-   strlcpy(s,t,nch+1);
-   char *sopen  = strchr(s,'<');
-   if (sopen == nullptr) {
-      Fatal("TStreamerSTL","For %s, the type name (%s) is seemingly not a template (template argument not found)", name, s);
+   if (arglist.fElements.size() < 2) {
+      Fatal("TStreamerSTL","For %s, the type name (%s) is seemingly not a template (template argument not found)", name, t);
       return;
    }
-   *sopen  = 0; sopen++;
-   // We are looking for the first arguments of the STL container, because
-   // this arguments can be a templates we need to count the < and >
-   char* current=sopen;
-   for(int count = 0; *current!='\0'; current++) {
-      if (*current=='<') count++;
-      if (*current=='>') {
-         if (count==0) break;
-         count--;
-      }
-      if (*current==',' && count==0) break;
-   }
-   char *sclose = current; *sclose = 0; sclose--;
-   char *sconst = strstr(sopen,"const ");
-   char *sbracket = strstr(sopen,"<");
-   if (sconst && (sbracket==nullptr || sconst < sbracket)) {
-      // the string "const" may be part of the classname!
-      char *pconst = sconst-1;
-      if (*pconst == ' ' || *pconst == '<' || *pconst == '*' || *pconst == '\0') sopen = sconst + 5;
-   }
-   fSTLtype = TClassEdit::STLKind(s);
-   fCtype   = 0;
-   if (fSTLtype == ROOT::kNotSTL) { delete [] s; return;}
-   if (dmPointer) fSTLtype += TVirtualStreamerInfo::kOffsetP;
 
-   // find STL contained type
-   while (*sopen==' ') sopen++;
-   Bool_t isPointer = kFALSE;
-   // Find stars outside of any template definitions in the
-   // first template argument.
-   char *star = strrchr(sopen,'>');
-   if (star) star = strchr(star,'*');
-   else star = strchr(sopen,'*');
-   if (star) {
-      isPointer = kTRUE;
-      *star = 0;
-      sclose = star - 1;
+   const std::string& inside_type = arglist.fElements[1];
+   std::string inside = (inside_type.find("const ")==0) ? inside_type.substr(6) : inside_type;
+
+   // Let's treat the unique_ptr case
+   bool nameChanged = false;
+   std::string intype = TClassEdit::GetNameForIO(inside, TClassEdit::EModType::kNone, &nameChanged);
+
+   bool isPointer = nameChanged; // unique_ptr is considered a pointer
```

Comment:
This is a bit implicit imho, I don't grasp fully the equivalence between changing the input type name and it being a pointer (is it only done for `unique_ptr`?). I would comment this further here

Reply:
`GetNameForIO` sole purpose (for now?) is to covert `unique_ptr<T>` to `T*` .... but it does so also inside template parameter and thus the shortcut is wrong.  Thanks for pointing me in the right direction.